doc/Plugin-API.md

```diff
@@ -66,12 +66,17 @@ To communicate with each other, the implementation of each side of the API - `Ma
 The proxy is based on the interface of the other side: `Main` implementation has a proxy of the `Ext` interface and vice versa.
 The implementations do not have explicit dependencies to each other.
 
-Communication via RPC only supports transferring plain JSON objects: Only pure DTO objects without any functions can be transmitted.
-Consequently, objects with functions need to be cached and references to such objects need to be transmitted as handles (ids) that are resolved to the original object later on to invoke functions.
+### Encoding and Decoding RPC Messages
 
-For instance, in [LanguagesExtImpl](https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/plugin/languages.ts)#registerCodeActionsProvider a new code action provider is cached on the `Ext` side and then registered on the `Main` side via its handle.
-When the code action provider’s methods are later invoked on the `Main` side (e.g. in [LanguagesMainImpl](https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/main/browser/languages-main.ts)#provideCodeActions), it calls the `Ext` side with this handle.
-The `Ext` side then gets the cached object, executes appropriate functions and returns the results back to the `Main` side (e.g. in [LanguagesExtImpl](https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/plugin/languages.ts)#$provideCodeActions).
+The communication between each side of the API is goverend by proxies that use a `RpcProtocol` on a given channel to transmit RPC messages such as requests and notifications.
+In Theia, the encoding and decoding process of RPC messages can be customized through dedicated `RpcMessageEncoder` and `RpcMessageDecoder` classes that can be provided when a new RpcProtocol is created.
+The `RpcMessageEncoder` writes RPC messages to a buffer whereas the `RpcMessageDecoder` parses a binary message from a buffer into a RPC message.
+
+By default, Theia uses an encoder and decoder based on [msgpackr](https://www.npmjs.com/package/msgpackr) that already properly handles many built-in types as well as arrays and maps.
+We can separately extend the encoding and decoding of our own classes by installing extensions using the `MsgPackExtensionManager` singleton, accessible from both ends of the channel.
```

Comment:
I used the word class here because that is what is used as key by msgpackr when adding an extension, see https://github.com/kriszyp/msgpackr?tab=readme-ov-file#custom-extensions. We probably could also use the word `type` here as what you really need to provide is a constructor (function) so that we can make an `instanceof` check (see https://github.com/kriszyp/msgpackr/blob/master/pack.js#L444).

With regard to your comment below, I'll try to discuss with other people what the best option is. I don't want to change it to "non-DTO object" if we actually would suggest to "replace" the usage of DTO objects with a custom serialization approach.

Reply:
Ah, ok. Makes sense to me.